### ISSUE TITLE
Core: Fuse Optimizer into Method

### ISSUE THREAD

**Issue by GPMueller** [spirit-code/spirit#279]
Using template specializations similar to https://stackoverflow.com/a/30440624/4069571, one could define different versions of a `Method::Solver_Step` function or something similar, discriminating between different solvers via e.g. `enum class Solver`.
The different optimization schemes could be implemented as specializations of the template method in single headers each, while the `Method` would be created e.g. as `Method_LLG<Solver::SIB>(...)`.

Example:
```C++
template <Solver solver, typename std::enable_if <solver == Solver::SIB> :: type* = nullptr> 
void Method::Solver_Step()
{
    this->Calculate_Force(...);
    // ... SIB temporal...
    this->Calculate_Force(...);
    // ... SIB final
    this->image = spins_final;
}
```

**Comment by GPMueller** [comment]
An alternative would certainly be to just give the `Method` base class the various `Solver_Step` functions in form of non-template functions such as `Step_SIB` and similar.
However, a templated version may be preferable, as then e.g. the following would be possible:
- call `this->Prepare_Solver<Solver::SIB>` in constructor
- call `this->Solver_Step<Solver::SIB>` in Iterate
where the `Prepare_Solver` function could allocate/initialize necessary arrays (which should only be done for those arrays that the corresp. solver will use, as e.g. SIB and NCG will use different temporaries).

**Comment by GPMueller** [comment]
It may also be an approach to write a `Simulation` base class which takes two template arguments, e.g. `Simulation<Method::LLG, Solver::SIB>(parameters)`.
The question then is how to cleanly separate the different Method and Solver functions from each-other (e.g. putting them in different files), without going against the typical C++ way.

**Comment by GPMueller** [comment]
Started implementing with b9d6fd55deee1aa5afe20f0f9f0599f14a8e80a9 on `feature-refactor-method`.
Basically completed the implementation with 135cc1fcb33f1452a65f2c44cd0aa3d36a3442f9.

A new `Method_Solver` template class is now derived from `Method`. It enables to cleanly separate the Solver implementations from the Methods which use Solvers and separate the Methods which do not use Solvers.

**Comment by GPMueller** [comment]
Merged into develop with 8402598137b1e30c01343542fa0d412cec59b340.
